includes/class-wcs-action-scheduler-customer-notifications.php

```diff
@@ -337,14 +337,16 @@ public function update_date( $subscription, $date_type, $datetime ) {
 	/**
 	 * Schedule notifications if the date has been deleted.
 	 *
-	 * @param object $subscription An instance of a WC_Subscription object
+	 * @param WC_Subscription $subscription An instance of a WC_Subscription object
 	 * @param string $date_type Can be 'trial_end', 'next_payment', 'end', 'end_of_prepaid_term' or a custom date type
 	 */
 	public function delete_date( $subscription, $date_type ) {
 		$action = $this->get_action_from_date_type( $date_type );
 		if ( $action ) {
 			$this->unschedule_actions( $action, self::get_action_args( $subscription ) );
 		}
+		//TODO: double check, if this is necessary.
+		$this->schedule_all_notifications( $subscription );
```

Comment:
Expanding a bit more here: 

I believe the process might run twice due to the hook at `woocommerce_before_subscription_object_save`, but if someone deletes a date _directly (e.g., calling `$subscription->delete_date()`)_, it should trigger the data store's `save_dates()` function, which will necessitate a rescheduling here.

This might also relate to what @james-allan described in his follow-up [list](https://github.com/woocommerce/woocommerce-subscriptions/issues/4737): _"Deleting a free trial date doesn’t schedule the pre-renewal email."_?